File: 238 product of array except self/ProductExceptSelf.java
class ProductExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int curProduct = 1;
        int[] res = new int[nums.length];
        res[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            curProduct *= nums[i - 1];
            res[i] = curProduct;
        }
        curProduct = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            res[i] *= curProduct;
            curProduct *= nums[i];
        }
        return res;
    }
}

File: 238 product of array except self/ProductExceptSelf.py
class ProductExceptSelf(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        res = [1] * n
        acc_product = nums[n - 1]
        
        for i in range(n - 2, -1, -1):
            res[i] = acc_product
            acc_product *= nums[i]
        
        acc_product = 1
        for i in range(n):
            res[i] = acc_product * res[i]
            acc_product *= nums[i]
        
        return res